File: src/interface/Python/paramonte/vis/kde2d.py
####################################################################################################################################
####################################################################################################################################
####
####   MIT License
####
####   ParaMonte: plain powerful parallel Monte Carlo library.
####
####   Copyright (C) 2012-present, The Computational Data Science Lab
####
####   This file is part of the ParaMonte library.
####
####   Permission is hereby granted, free of charge, to any person obtaining a 
####   copy of this software and associated documentation files (the "Software"), 
####   to deal in the Software without restriction, including without limitation 
####   the rights to use, copy, modify, merge, publish, distribute, sublicense, 
####   and/or sell copies of the Software, and to permit persons to whom the 
####   Software is furnished to do so, subject to the following conditions:
####
####   The above copyright notice and this permission notice shall be 
####   included in all copies or substantial portions of the Software.
####
####   THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, 
####   EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF 
####   MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. 
####   IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, 
####   DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR 
####   OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE 
####   OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
####
####   ACKNOWLEDGMENT
####
####   ParaMonte is an honor-ware and its currency is acknowledgment and citations.
####   As per the ParaMonte library license agreement terms, if you use any parts of 
####   this library for any purposes, kindly acknowledge the use of ParaMonte in your 
####   work (education/research/industry/development/...) by citing the ParaMonte 
####   library as described on this page:
####
####       https://github.com/cdslaborg/paramonte/blob/master/ACKNOWLEDGMENT.md
####
####################################################################################################################################
####################################################################################################################################

import numpy as np
from scipy.fft import dctn, idctn
from scipy.optimize import brentq

####################################################################################################################################
#### kde2d
####################################################################################################################################

def kde2d(x, y, n=256, limits=None):
    """

    Return the 2d density map from discrete observations via 
    2-dimensional diffusion Kernel density estimation.

    First the input data is binned. After binning, the function 
    determines the optimal bandwidth according to the diffusion-based 
    method. It then smooths the binned data over the grid using a 
    Gaussian kernel with a standard deviation corresponding to 
    that bandwidth.

        This module is based on the KDE estimation method of 

            Z. I. Botev, J. F. Grotowski, D. P. Kroese: 
            Kernel density estimation via diffusion. 
            Annals of Statistics 38 (2010), no. 5, 2916--2957. 
            doi:10.1214/10-AOS799

        and 

            John Hennig 
            DOI: 10.5281/zenodo.3830437 

        **Parameters**

            x

                A lists of array of numbers that represent discrete 
                observations of a random variable with two coordinate 
                components. The observations are binned on a grid of 
                n*n points, where ``n`` must be a power of 2 or will 
                be coerced to the next one. If ``x`` and ``y`` are not 
                the same length, the algorithm will raise a ``ValueError``.

            y

                A lists of array of numbers that represent discrete 
                observations of a random variable with two coordinate 
                components. The observations are binned on a grid of 
                n*n points, where ``n`` must be a power of 2 or will 
                be coerced to the next one. If ``x`` and ``y`` are not 
                the same length, the algorithm will raise a ``ValueError``.

            n (optional)

                The number of grid points. It must be a power of 2. 
                Otherwise, it will be coerced to the next power of two.
                The default is 256.

            limits (optional)

                Data ``limits`` specified as a tuple of tuples denoting
                ``((xmin, xmax), (ymin, ymax))``. If any of the values 
                are ``None``, they will be inferred from the data. Each 
                tuple, or even both of them, may also be replaced by a 
                single value denoting the upper bound of a range 
                centered at zero. The default is ``None``.

        **Returns**

            A tuple whose elements are the following:

                density

                    The density map of the data.

                grid

                    The grid at which the density is computed.

                bandwidth

                    The optimal values (per axis) that the algorithm has 
                    determined. If the algorithm does not converge, it 
                    will raise a ``ValueError``.

    ---------------------------------------------------------------------------
    """

    # Convert to arrays in case lists are passed in.

    x = np.array(x)
    y = np.array(y)

    # Make sure numbers of data points are consistent.

    N = len(x)
    if len(y) != N:
        raise ValueError('x and y must have the same length.')

    # Round up number of bins to next power of two.

    n = int(2**np.ceil(np.log2(n)))

    # Determine missing data limits.

    if limits is None:

        xmin = xmax = ymin = ymax = None

    elif isinstance(limits, tuple):

        (xlimits, ylimits) = limits
        if xlimits is None:
            xmin = xmax = None
        elif isinstance(xlimits, tuple):
            (xmin, xmax) = xlimits
        else:
            xmin = -xlimits
            xmax = +xlimits
        if ylimits is None:
            ymin = ymax = None
        elif isinstance(ylimits, tuple):
            (ymin, ymax) = ylimits
        else:
            ymin = -ylimits
            ymax = +ylimits
    else:

        xmin = -limits
        xmax = +limits
        ymin = -limits
        ymax = +limits

    if None in (xmin, xmax):
        delta = x.max() - x.min()
        if xmin is None:
            xmin = x.min() - delta/4
        if xmax is None:
            xmax = x.max() + delta/4

    if None in (ymin, ymax):
        delta = y.max() - y.min()
        if ymin is None:
            ymin = y.min() - delta/4
        if ymax is None:
            ymax = y.max() + delta/4

    deltax = xmax - xmin
    deltay = ymax - ymin

    # Bin samples on regular grid.

    (binned, xedges, yedges) = np.histogram2d(x, y, bins=n, range=((xmin, xmax), (ymin, ymax)))
    grid = (xedges[:-1], yedges[:-1])

    # Compute discrete cosine transform. Adjust first component.

    transformed = dctn(binned/N)
    transformed[0, :] /= 2
    transformed[:, 0] /= 2

    # Pre-compute squared indices and transform components before solver loop.

    k  = np.arange(n, dtype="float") # float avoids integer overflow.
    k2 = k**2
    a2 = transformed**2

    # Define internal functions to be solved iteratively.

    def γ(t):
        sigma = psi(0, 2, t) + psi(2, 0, t) + 2*psi(1, 1, t)
        γ = (2*np.pi*N*sigma)**(-1/3)
        return (t - γ) / γ

    def psi(i, j, t):
        if i + j <= 4:
            sigma  = abs(psi(i+1, j, t) + psi(i, j+1, t))
            C  = (1 + 1/2**(i+j+1)) / 3
            pii = np.product(np.arange(1, 2*i, 2))
            pij = np.product(np.arange(1, 2*j, 2))
            t  = (C*pii*pij / (np.pi*N*sigma)) ** (1/(2+i+j))
        w = 0.5 * np.ones(n)
        w[0] = 1
        w = w * np.exp(-np.pi**2 * k2*t)
        wx = w * k2**i
        wy = w * k2**j
        return (-1)**(i+j) * np.pi**(2*(i+j)) * wy @ a2 @ wx

    # Solve for optimal diffusion time t*.

    try:
        ts = brentq(lambda t: t - γ(t), 0, 0.1)
    except ValueError:
        raise ValueError('Bandwidth optimization did not converge.') from None

    # Calculate diffusion times along x- and y-axis.

    psi02 = psi(0, 2, ts)
    psi20 = psi(2, 0, ts)
    psi11 = psi(1, 1, ts)
    tx1 = (psi02**(3/4) / (4*np.pi*N*psi20**(3/4) * (psi11 + np.sqrt(psi02*psi20))) )**(1/3)
    tx2 = (psi20**(3/4) / (4*np.pi*N*psi02**(3/4) * (psi11 + np.sqrt(psi02*psi20))) )**(1/3)

    # Note:
    # The above uses the nomenclature from the paper. In the Matlab
    # reference, tx1 is called t_y, while tx2 is t_x. This is a curious
    # change in notation. It may be related to the fact that image
    # coordinates are typically in (y,x) index order, whereas matrices,
    # such as the binned histogram (in Matlab as much as in Python),
    # are in (x,y) order. The Matlab code eventually does return
    # image-like index order, though it never explicitly transposes
    # the density matrix. That is implicitly handled by its custom
    # implementation of the inverse transformation (idct2d), which
    # only employs one matrix transposition, not two as its forward
    # counterpart (dct2d).

    # Apply Gaussian filter with optimized kernel.

    smoothed = transformed * np.outer(np.exp(-np.pi**2 * k2 * tx2/2), np.exp(-np.pi**2 * k2 * tx1/2))

    # Reverse transformation.

    smoothed[0, :] *= 2
    smoothed[:, 0] *= 2
    inverse = idctn(smoothed)

    # Normalize density.

    density = np.transpose(inverse) * n/deltax * n/deltay

    # Determine bandwidth from diffusion times.

    bandwidth = np.array([np.sqrt(tx2)*deltax, np.sqrt(tx1)*deltay])

    # Return results.

    return (density, grid, bandwidth)

File: src/interface/MATLAB/paramonte/vis/kde2d.m
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%%%
%%%%   MIT License
%%%%
%%%%   ParaMonte: plain powerful parallel Monte Carlo library.
%%%%
%%%%   Copyright (C) 2012-present, The Computational Data Science Lab
%%%%
%%%%   This file is part of the ParaMonte library.
%%%%
%%%%   Permission is hereby granted, free of charge, to any person obtaining a 
%%%%   copy of this software and associated documentation files (the "Software"), 
%%%%   to deal in the Software without restriction, including without limitation 
%%%%   the rights to use, copy, modify, merge, publish, distribute, sublicense, 
%%%%   and/or sell copies of the Software, and to permit persons to whom the 
%%%%   Software is furnished to do so, subject to the following conditions:
%%%%
%%%%   The above copyright notice and this permission notice shall be 
%%%%   included in all copies or substantial portions of the Software.
%%%%
%%%%   THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, 
%%%%   EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF 
%%%%   MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. 
%%%%   IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, 
%%%%   DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR 
%%%%   OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE 
%%%%   OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
%%%%
%%%%   ACKNOWLEDGMENT
%%%%
%%%%   ParaMonte is an honor-ware and its currency is acknowledgment and citations.
%%%%   As per the ParaMonte library license agreement terms, if you use any parts of 
%%%%   this library for any purposes, kindly acknowledge the use of ParaMonte in your 
%%%%   work (education/research/industry/development/...) by citing the ParaMonte 
%%%%   library as described on this page:
%%%%
%%%%       https://github.com/cdslaborg/paramonte/blob/master/ACKNOWLEDGMENT.md
%%%%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%
% Copyright (c) 2015, Dr. Zdravko Botev
% All rights reserved.
% 
% Redistribution and use in source and binary forms, with or without
% modification, are permitted provided that the following conditions are
% met:
% 
%     * Redistributions of source code must retain the above copyright
%       notice, this list of conditions and the following disclaimer.
%     * Redistributions in binary form must reproduce the above copyright
%       notice, this list of conditions and the following disclaimer in
%       the documentation and/or other materials provided with the distribution
%     * Neither the name of the The University of New South Wales nor the names
%       of its contributors may be used to endorse or promote products derived
%       from this software without specific prior written permission.
% 
% THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
% AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
% IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
% ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE
% LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
% CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
% SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
% INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
% CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
% ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
% POSSIBILITY OF SUCH DAMAGE.
% 
function [bandwidth,density,X,Y]=kde2d(data,n,MIN_XY,MAX_XY)
    % fast and accurate state-of-the-art
    % bivariate kernel density estimator
    % with diagonal bandwidth matrix.
    % The kernel is assumed to be Gaussian.
    % The two bandwidth parameters are
    % chosen optimally without ever
    % using/assuming a parametric model for the data or any "rules of thumb".
    % Unlike many other procedures, this one
    % is immune to accuracy failures in the estimation of
    % multimodal densities with widely separated modes (see examples).
    % INPUTS: data - an N by 2 array with continuous data
    %            n - size of the n by n grid over which the density is computed
    %                n has to be a power of 2, otherwise n=2^ceil(log2(n));
    %                the default value is 2^8;
    % MIN_XY,MAX_XY- limits of the bounding box over which the density is computed;
    %                the format is:
    %                MIN_XY=[lower_Xlim,lower_Ylim]
    %                MAX_XY=[upper_Xlim,upper_Ylim].
    %                The dafault limits are computed as:
    %                MAX=max(data,[],1); MIN=min(data,[],1); Range=MAX-MIN;
    %                MAX_XY=MAX+Range/4; MIN_XY=MIN-Range/4;
    % OUTPUT: bandwidth - a row vector with the two optimal
    %                     bandwidths for a bivaroate Gaussian kernel;
    %                     the format is:
    %                     bandwidth=[bandwidth_X, bandwidth_Y];
    %          density  - an n by n matrix containing the density values over the n by n grid;
    %                     density is not computed unless the function is asked for such an output;
    %              X,Y  - the meshgrid over which the variable "density" has been computed;
    %                     the intended usage is as follows:
    %                     surf(X,Y,density)
    % Example (simple Gaussian mixture)
    % clear all
    %   % generate a Gaussian mixture with distant modes
    %   data=[randn(500,2);
    %       randn(500,1)+3.5, randn(500,1);];
    %   % call the routine
    %     [bandwidth,density,X,Y]=kde2d(data);
    %   % plot the data and the density estimate
    %     contour3(X,Y,density,50), hold on
    %     plot(data(:,1),data(:,2),'r.','MarkerSize',5)
    %
    % Example (Gaussian mixture with distant modes):
    %
    % clear all
    %  % generate a Gaussian mixture with distant modes
    %  data=[randn(100,1), randn(100,1)/4;
    %      randn(100,1)+18, randn(100,1);
    %      randn(100,1)+15, randn(100,1)/2-18;];
    %  % call the routine
    %    [bandwidth,density,X,Y]=kde2d(data);
    %  % plot the data and the density estimate
    %  surf(X,Y,density,'LineStyle','none'), view([0,60])
    %  colormap hot, hold on, alpha(.8)
    %  set(gca, 'color', 'blue');
    %  plot(data(:,1),data(:,2),'w.','MarkerSize',5)
    %
    % Example (Sinusoidal density):
    %
    % clear all
    %   X=rand(1000,1); Y=sin(X*10*pi)+randn(size(X))/3; data=[X,Y];
    %  % apply routine
    %  [bandwidth,density,X,Y]=kde2d(data);
    %  % plot the data and the density estimate
    %  surf(X,Y,density,'LineStyle','none'), view([0,70])
    %  colormap hot, hold on, alpha(.8)
    %  set(gca, 'color', 'blue');
    %  plot(data(:,1),data(:,2),'w.','MarkerSize',5)
    %
    %  Reference:
    % Kernel density estimation via diffusion
    % Z. I. Botev, J. F. Grotowski, and D. P. Kroese (2010)
    % Annals of Statistics, Volume 38, Number 5, pages 2916-2957.
    global N A2 I
    if nargin<2
        n=2^8;
    end
    n=2^ceil(log2(n)); % round up n to the next power of 2;
    N=size(data,1);
    if nargin<3
        MAX=max(data,[],1); MIN=min(data,[],1); Range=MAX-MIN;
        MAX_XY=MAX+Range/2; MIN_XY=MIN-Range/2;
    end
    scaling=MAX_XY-MIN_XY;
    if N<=size(data,2)
        error('data has to be an N by 2 array where each row represents a two dimensional observation')
    end
    transformed_data=(data-repmat(MIN_XY,N,1))./repmat(scaling,N,1);
    %bin the data uniformly using regular grid;
    initial_data=ndhist(transformed_data,n);
    % discrete cosine transform of initial data
    a= dct2d(initial_data);
    % now compute the optimal bandwidth^2
      I=(0:n-1).^2; A2=a.^2;
     t_star=root(@(t)(t-evolve(t)),N);
    p_02=func([0,2],t_star);p_20=func([2,0],t_star); p_11=func([1,1],t_star);
    t_y=(p_02^(3/4)/(4*pi*N*p_20^(3/4)*(p_11+sqrt(p_20*p_02))))^(1/3);
    t_x=(p_20^(3/4)/(4*pi*N*p_02^(3/4)*(p_11+sqrt(p_20*p_02))))^(1/3);
    % smooth the discrete cosine transform of initial data using t_star
    a_t=exp(-(0:n-1)'.^2*pi^2*t_x/2)*exp(-(0:n-1).^2*pi^2*t_y/2).*a; 
    % now apply the inverse discrete cosine transform
    if nargout>1
        density=idct2d(a_t)*(numel(a_t)/prod(scaling));
        density(density<0)=eps; % remove any negative density values
        [X,Y]=meshgrid(MIN_XY(1):scaling(1)/(n-1):MAX_XY(1),MIN_XY(2):scaling(2)/(n-1):MAX_XY(2));
    end
    bandwidth=sqrt([t_x,t_y]).*scaling; 
end

%#######################################

function [out,time]=evolve(t)
    global N
    Sum_func = func([0,2],t) + func([2,0],t) + 2*func([1,1],t);
    time=(2*pi*N*Sum_func)^(-1/3);
    out=(t-time)/time;
end

%#######################################

function out=func(s,t)
    global N
    if sum(s)<=4
        Sum_func=func([s(1)+1,s(2)],t)+func([s(1),s(2)+1],t); const=(1+1/2^(sum(s)+1))/3;
        time=(-2*const*K(s(1))*K(s(2))/N/Sum_func)^(1/(2+sum(s)));
        out=psi(s,time);
    else
        out=psi(s,t);
    end
end

%#######################################

function out=psi(s,Time)
    global I A2
    % s is a vector
    w=exp(-I*pi^2*Time).*[1,.5*ones(1,length(I)-1)];
    wx=w.*(I.^s(1));
    wy=w.*(I.^s(2));
    out=(-1)^sum(s)*(wy*A2*wx')*pi^(2*sum(s));
end

%#######################################

function out=K(s)
    out=(-1)^s*prod((1:2:2*s-1))/sqrt(2*pi);
end

%#######################################

function data=dct2d(data)
    % computes the 2 dimensional discrete cosine transform of data
    % data is an nd cube
    [nrows,ncols]= size(data);
    if nrows~=ncols
        error('data is not a square array!')
    end
    % Compute weights to multiply DFT coefficients
    w = [1;2*(exp(-i*(1:nrows-1)*pi/(2*nrows))).'];
    weight=w(:,ones(1,ncols));
    data=dct1d(dct1d(data)')';
    function transform1d=dct1d(x)

        % Re-order the elements of the columns of x
        x = [ x(1:2:end,:); x(end:-2:2,:) ];

        % Multiply FFT by weights:
        transform1d = real(weight.* fft(x));
    end
end

%#######################################

function data = idct2d(data)
    % computes the 2 dimensional inverse discrete cosine transform
    [nrows,ncols]=size(data);
    % Compute wieghts
    w = exp(i*(0:nrows-1)*pi/(2*nrows)).';
    weights=w(:,ones(1,ncols));
    data=idct1d(idct1d(data)');
    function out=idct1d(x)
        y = real(ifft(weights.*x));
        out = zeros(nrows,ncols);
        out(1:2:nrows,:) = y(1:nrows/2,:);
        out(2:2:nrows,:) = y(nrows:-1:nrows/2+1,:);
    end
end

%#######################################

function binned_data=ndhist(data,M)
    % this function computes the histogram
    % of an n-dimensional data set;
    % 'data' is nrows by n columns
    % M is the number of bins used in each dimension
    % so that 'binned_data' is a hypercube with
    % size length equal to M;
    [nrows,ncols]=size(data);
    bins=zeros(nrows,ncols);
    for i=1:ncols
        [dum,bins(:,i)] = histc(data(:,i),[0:1/M:1],1);
        bins(:,i) = min(bins(:,i),M);
    end
    % Combine the  vectors of 1D bin counts into a grid of nD bin
    % counts.
    binned_data = accumarray(bins(all(bins>0,2),:),1/nrows,M(ones(1,ncols)));
end

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

function t=root(f,N)
    % try to find smallest root whenever there is more than one
    N=50*(N<=50)+1050*(N>=1050)+N*((N<1050)&(N>50));
    tol=10^-12+0.01*(N-50)/1000;
    flag=0;
    while flag==0
        try
            t=fzero(f,[0,tol]);
            flag=1;
        catch
            tol=min(tol*2,.1); % double search interval
        end
        if tol==.1 % if all else fails
            t=fminbnd(@(x)abs(f(x)),0,.1); flag=1;
        end
    end
end
